Adding the  Off-Path Packet Forwarding text and the no_op message (#12)
tlswg/dtls-rrc: draft-ietf-tls-dtls-rrc.md

# Off-Path Packet Forwarding {#off-path}

An off-path attacker that can observe packets might forward copies of
genuine packets to endpoints. If the copied packet arrives before
the genuine packet, this will appear as a NAT rebinding. Any genuine
packet will be discarded as a duplicate. If the attacker is able to
continue forwarding packets, it might be able to cause migration to a
path via the attacker. This places the attacker on-path, giving it
the ability to observe or drop all subsequent packets.

This style of attack relies on the attacker using a path that has
approximately the same characteristics as the direct path between
endpoints. The attack is more reliable if relatively few packets are
sent or if packet loss coincides with the attempted attack.

A data packet received on the original path that increases the
maximum received packet number will cause the endpoint to move back
to that path. Eliciting packets on this path increases the
likelihood that the attack is unsuccessful.

{{fig-off-path}} demonstrates the case where a receiver receives a
packet with a new source IP address and/or new port number. The
receiver needs to determine whether this path change is caused

> by an attacker and will sent a RRC message with a path_challenge
> along the old path. 

Reviewer: ```suggestion
by an attacker and will send a RRC message of type path_challenge (RRC-1)
on the old path. 
```

~~~~
        new   +--------+  old
        path  |        |  path
       +----->|Receiver|<-----+
       |      |        |      |
       |      +--------+      |
       |                      |
       |                      |
       |                      |
       |                      |
       |                      |
 +----------+                 |
 | Attacker?|                 |
 +----------+                 |
       |                      |
       |                      |
       |                      |
       |      +--------+      |
       |      |        |      |
       +------| Sender |------+
              |        |
              +--------+
~~~~
{: #fig-off-path title="Off-Path Packet Forwarding Scenario"}

Three cases need to be considered:

Case 1: The old path is dead, which leads to a timeout of RRC-1. 

> As shown in {{fig-old-path-dead}}, RRC-2 needs to be sent on the

Reviewer: ```suggestion
As shown in {{fig-old-path-dead}}, a RRC message of type path_challenge (RRC-2) needs to be sent on the
```

new path. In this situation the switch to the new path is

> considered legitimate (assuming that the attacker is unable to

Reviewer: I'd remove the parenthetical because this is the overall assumption not something that is specific to this situation.

  .      with *|                      |* with       .
  .     path- *|                      |* path-      .
  . challenge *|                      |* challenge  .
  .           *|                      |*            .
  .           *|                      |*            .
  .      +----------+                 |*            .
  .      | Attacker |                 |*            .
  .      +----------+                 |*            .
  .           *|                      |*            .
  .           *|                      |*            .
  .           *|                      |*            .
  .RRC-4      *|      +--------+      |*       RRC-2.
  .with       *|      |        |      |*        with.
  .path-      *+------| Sender |------+*       path-.
  .response   *******>|        |<*******      delete.
  ....................+--------+.....................
~~~~
{: #fig-old-path-not-preferred title="Old path is not preferred"}

Case 3: The old path is alive and preferred. 

This is most likely the result of an attacker. The sender replies 
with RRC-2 containing a path_response along the old path. The
interaction is shown in {{fig-old-path-preferred}}. This results

> in the connection being migrate back to the old path.

Reviewer: ```suggestion
in the connection being migrated back to the old path.
```

The `return_routability_check` message MUST be authenticated and encrypted using
the currently active security context.

# Off-Path Packet Forwarding {#off-path}

> An off-path attacker that can observe packets might forward copies of
> genuine packets to endpoints. If the copied packet arrives before
> the genuine packet, this will appear as a NAT rebinding. Any genuine
> packet will be discarded as a duplicate. If the attacker is able to
> continue forwarding packets, it might be able to cause migration to a
> path via the attacker. This places the attacker on-path, giving it
> the ability to observe or drop all subsequent packets.
> 
> This style of attack relies on the attacker using a path that has
> approximately the same characteristics as the direct path between
> endpoints. The attack is more reliable if relatively few packets are
> sent or if packet loss coincides with the attempted attack.
> 
> A non-path probing packet received on the original path that increases the
> maximum received packet number will cause the endpoint to move back
> to that path. Eliciting packets on this path increases the
> likelihood that the attack is unsuccessful. Therefore, mitigation of
> this attack relies on triggering the exchange of packets.

Reviewer: I guess we can replace this with a reference to the relevant section of QUIC?
```suggestion
The same off-path attack described in {{Section 9.3.3 of RFC9000}} applies to DTLS sessions that use connection identifiers.
```

> In response to an apparent migration, endpoints MUST validate the

Reviewer: Is this a MUST or a SHOULD?  If it's a MUST it's probably better to describe along  with the path validation algorithm?

previously active path using a `return_routability_check`  exchange.
This induces the sending of new packets on that path. If the path is
no longer viable, the validation attempt will time out and fail;
if the path is viable but no longer desired, the validation will

> succeed but only results in probing packets being sent on the path.

Reviewer: we should define what we mean by "probing packet"

An endpoint that receives a `return_routability_check` message
containing a path_challenge on an active path SHOULD send 
a `return_routability_check` of msg-type no_op in response. 
If the `return_routability_check` with a `no_op` arrives before any
copy made by an attacker, this results in the connection being
migrate back to the original path. Any subsequent migration to

> another path restarts this entire process.

Reviewer: I am probably misunderstanding the mechanics here.  But I'd have thought that if I receive a path_challenge on a path that is active but I don't want to use any more, I should state that explicitly using no_op (which I'd rather call path_drop)?

Note that this defense is imperfect, but this is not considered a serious
problem. If the path via the attack is reliably faster than the
original path despite multiple attempts to use that original path, it
is not possible to distinguish between an attack and an improvement
in routing.

An endpoint could also use heuristics to improve detection of this
style of attack. For instance, NAT rebinding is improbable if
packets were recently received on the old path; similarly, rebinding
is rare on IPv6 paths. Endpoints can also look for duplicated

> packets. Conversely, a change in connection ID is more likely to

Author: (Add remark that there is no change in connection ID with DTLS 1.2)

{{path-challenge-reqs}} and {{path-response-reqs}} contain the requirements for
the initiator and responder roles, broken down per protocol phase.

# Enhanced Path Validation Procedure

Note: This algorithm also takes the {{off-path}} scenario into account.

The receiver that observes the peer's address or port update MUST stop sending
any buffered application data (or limit the data sent to the unvalidated
address to the anti-amplification limit) and initiate the return routability
check that proceeds as follows:

1. An unpredictable cookie is placed in a `return_routability_check` message of
   type path_challenge;
1. The message is sent to the previously valid address and a timer T (see
   {{timer-choice}}) is started;
1. The peer endpoint, after successfully verifying the received
   `return_routability_check` message responds by echoing the cookie value in a
   `return_routability_check` message of type path_response if the path through
   which the message was received is preferred. If the path is not preferred,
   a `return_routability_check` message of type path_delete MUST be returned;

> 1. When the initiator receives and verifies the `return_routability_check`

Reviewer: ```suggestion
1. When the initiator receives and verifies that the `return_routability_check`
```

>    message contains the sent cookie, it updates the peer address binding;

Reviewer: the "updates the peer address binding" is for when the receiver gets back a verified path_response.  However, at this point it could have received a path_delete, and that'd produce a different action.

## Path Response Requirements {#path-response-reqs}

* The responder MUST NOT delay sending an elicited path_response or
  path_delete messages.
* The responder MUST send exactly one path_response or path_delete message 

>   for each received path_request.

Reviewer: ```suggestion
  for each received path_challenge.
```
